src/mono/mono/mini/method-to-ir.c

```diff
@@ -5850,7 +5850,8 @@ handle_ctor_call (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fs
 
 	/* Avoid virtual calls to ctors if possible */
 	if (!context_used && !rgctx_arg) {
-		INLINE_FAILURE ("inline failure");
+		if (!m_method_is_aggressive_inlining (cfg->current_method) && !m_method_is_aggressive_inlining (cmethod))
+			INLINE_FAILURE ("ctor call");
```

Comment:
@vargaz, why only if they are marked aggressive inlining rather than in general?

There are a lot of methods that are small and call constructors or other things that aren't marked this way that would benefit.

Consider for example `op_Explicit` and `op_Implicit` cast operators that are implemented simply as:
```csharp
public static implicit operator BigInteger(byte value) => new BigInteger(value);
```

We likewise have constructors that defer to other constructors and similar where many of these are very small methods that basically just forward to others.

Reply:
The mono inlining heuristics are not very good, this PR just fixes an issue where inlining was disabled even through the callee had the aggressive inlining flag.

Reply:
Is there an issue tracking Mono improving these heuristics?

We (the libraries team) don't typically put `AggressiveInlining` on small methods, particularly methods `<= 16 bytes of IL` -- This is the "always inline threshold" for RyuJIT, barring cases like PGO or the method always throwing, which would indicate its a likely cold path.

We instead tend to only put it on methods which are known to be extra hot or known to be problematic for the JIT, such as having a large amount of code that can be elided as part of dead code optimizations, generic specialization, etc.

Reply:
(so this may be a general place where Mono could see some easy perf wins by adjusting its heuristics and becoming more similar to what RyuJIT is doing)

Reply:
Its possible. In this particular case, INLINE_FAILURE means we abort inlining the current method if it makes a call to a ctor.